Character.cpp

```diff
@@ -89,6 +88,16 @@ Character Character::parseUnit(const std::string& FileName)
 	return Character(name, hp, dmg);
 }
 
+void Player::DMGTaken(const Character character)
+{
+	this->setXp();
+	if (this->getXp() >= 100) {
+		this->LevelUp();
+	}
+	hp -= character.getDmg();
+	if (hp <= 0) { hp = 0; }
+}
```

Comment:
Hmm, jól értelmezem, hogy most az kapja az xp-t, akit megsebeztek? Értem én, hogy ami nem öl meg, az erősebbé tesz, de a játékokban sem így szokott lenni, meg a feladatleírásban sem ez volt megadva.
Szóval a támadó kapjon xp-t, és arra is figyelj, hogy csak annyit kapjon, amennyi sérülést ténylegesen okozott. Ez lehet kevesebb, mint a teljes sebzés, amire egy ütésből képes.

Reply:
Ui.: Megnéztem a feladatleírást. Némileg félreérthető lehet az utolsó pont, miszerint: "az XP novelo logika fusson le akkor is, ha a playert tamadjak meg, nem csak akkor, ha O tamad". Itt a "támad" az első ütésre utal, tehát ha a másik visszaüt, akkor ő is kapjon xp-t.

Reply:
@oliverosz Kicsit én is belekavarodtam most. Szóval úgy tudom hogy van egy Player-ünk és csakis ő kap xp-t, akivel harcol az nem, az csak egy átlagos ellenfél. A Player viszont mivel xp-t kap ugye leveluppol és erősödik. Szóval itt elvileg mégis jól működik ez a DMGTaken függvény mert csak a Player osztálynak van xp változója és a Fight függvényünk egy Player típusú egységet harcoltat egy Character típusúval. Mivel a Character típusú egyszerű harcosnak nincs is xp-je, lehetetlen hogy ő szintet lépne. Azonban pl mikor Luke csap össze Vaderrel, Luke a Player és 250 a kezdő hp-ja, de 306-al győzi le Vadert. A tesztek során ki is írattam a harc lépéseit és az egyszerű harcos (mint az előbb Vader) hp-ja csak csökkent és a damage értéke stagnált, így úgy gondolom lehetetlen hogy a rossz harcos kapna xp-t és erősödne. Kérlek mond el hol hibázhatok vagy gondolom rosszul de úgy vélem jól működik a program, mert a DMGTaken-ben az kapja az xp-t aki ténylegesen üt, csak kicsit kesze-kuszán írtuk meg, tudom nagyon furának hat így ránézésre valóban!

Reply:
@Szabi1104 Nem az a baj, hogy a Character szintet lép (ilyen tényleg nem fordul elő), hanem hogy a Player akkor kap xp-t, amikor őt ütik, nem akkor, amikor ő üt. A Player a feladat szerint "minden utes utan kap annyi XP-t, amennyi sebzest bevitt". Itt viszont ő sérül, és nem ő visz be sebzést, mégis kap xp-t, és ez a hibás. Amikor meg ő üt, akkor nem kap xp-t. És így az xp mennyisége se jó, mert a Player nem annyi xp-t kap, amennyi sebzést bevitt, hanem amennyit elszenvedett.

Reply:
@oliverosz Az xp mennyiségének a hibáját ki tudom javítani. Azonban sajnos nem tudok a másik, Character osztályhoz tartozó DMGTaken függvényben xp-t adni mert ugye ennek a szülő osztálynak nincs xp tagja. Sajnos ehhez a feladathoz a mi eddigi logikánk nagyon is bakavar, mert nálunk nem két olyan függvény van ami az adott fél ütését végzi, hanem két olyan ami azt kezelik le mikor az adott felet megütik, annak életereje az ütő (másik fél) sebzésével csökkenjen (min. 0-ra). Szóval légyszi mond meg hogy a mi logikánknál hogy tehetnék xp-t a Player (a hős) DMGTaken függvényébe, vagy milyen megoldás lenne lehetséges? Mert a korábbi verzió működött, de így xp-t nem írhatok a szülőbe (Character), viszont mégis nála, az ő függvényében kellene mert az ő DMGTaken-je kezeli le a Player ütését. 
Köszönöm előre is a segítségedet, hogyan csinálhatnám mindezt!

[Megjegyzés: Értem hogy így csúnya de egyébként alapvetően nincs gond szerintem ezzel a logikával sem, mert Player üt, igen ott csak sebzi Character-t, ha megöli akkor már tárgytalan az xp és a szintlépés (úgy gondolom) mert vége a harcnak. Ezután ha életben van következik Character aki viszont mielőtt ütne, bepótoljuk ami Player-nél lemaradt és Player előbb az előbbi ütéséért xp-t kap (tudom, még nem teljesen jól, valóban) és akár szintet lép. Szóval mindez megtörténik rögtön az ütése után lényegében csak (tudom csúnyán) de ebben a másik függvényben. Aztán kapja csak be Character ütését. Szóval értem ez így nem szép, de elvileg jól működik, persze csak amennyiben ha Player győz mert végez Characterrel, ott már nem fontos az xp kapás. Köszönöm előre is a véleményed és a javaslatodat, hogy ha ez így semmiképp nem maradhatna akkor esetleg lehetne-e az előző állapot vagy hogyan tudnám orvosolni a problémát!]

Reply:
> Azonban sajnos nem tudok a másik, Character osztályhoz tartozó DMGTaken függvényben xp-t adni mert ugye ennek a szülő osztálynak nincs xp tagja.

De nem is a Characternek kellene xp-t adni, hanem az őt támadó Playernek. Csak ehhez jelen verzióban a Characternek ismernie kellene a Player működését, ami ellentmond annak, hogy a Player a Character leszármazottja. Egy osztálynak nem szabad tudnia a leszármazottairól.

Szóval a `Character::DMGTaken` nem képes lekezelni, hogy a támadó xp-t kapjon. Meg kell fordítani a logikát: a támadónak legyen egy `attack` metódusa, amivel meg tud ütni egy Charactert. És ha ez virtuális függvény, akkor ez már attól függően működhet másként, hogy a támadó egy Player vagy egy Character. Előbbinél az xp-t is növelje meg a bevitt sebzéssel.

Reply:
@oliverosz Rendben van, értem. Azonban nem lesz gond ha így megszüntetem az eddigi DMGTaken függvényeket és helyettük az Attack lesz, így a későbbiekben a Json parser és Attackspeed-el való egyesítéskor mi fog történni ha nálam már nincs DMGTaken és más lett a logika is alapvetően?
Köszönöm előre is a válaszod!

Reply:
@oliverosz Vagy maradjon meg a DMGTaken ezzel a sebzési módszerrel, csak hozzá kellene adni ezt az attack metódust ami Player támadó esetén ad neki xp-t? Szeretnék még ahhoz segítséget kérni hogy tudnám megnézni itt hogy Player-e vagy Character a támadó ugye. Köszönöm előre is a válaszod és a segítséget!

Reply:
@Szabi1104 Nem baj, hogy megszünteted a DMGTakent. Az attackspeeddel így is, úgy is lesz conflict, amit az egyesítésnél fel kell oldani. Ezt megmutattam a gites videóban is.
Dynamic casttal ugyan meg tudná nézni a Character::DMGTaken, hogy a támadó Character vagy Player, de mint mondtam, ez rossz tervezési megoldás, az ősosztálynak nem szabad tudnia a leszármazottairól.

Reply:
@oliverosz Rendben, akkor igyekszem minél előbb úgy átírni a kódot ahogy mondtad.